### PR TITLE
Kill existing matching processes on startup

### PR DESCRIPTION
**Pull request by manics**
I thought parsing `/proc/*/cmdline` was simpler than installing psutil.

I've only tested this in a VM (following https://github.com/yuvipanda/cryptnono/pull/10), not as a K8s container.

In one terminal run
```sh
sh -c 'sleep 1h; /bin/echo cryptnono.banned.string1'
```
In another run
```
sudo ./scripts/execwhacker.py --config example/config.json --debug
```

### CODE REVIEW

**Comment by manics** [comment]
On second thoughts this probably requires mounting `/proc` into the container....

**Comment by manics** [comment]
Not tested on k8s, but this seems to work on a VM: `sudo podman run -it --rm --privileged --pid=host -v /usr/src:/usr/src:ro -v /sys:/sys:ro -v /lib/modules:/lib/modules:ro -v $PWD/example:/example -v /proc:/proc:ro docker.io/manics/cryptnono:pr11 /scripts/execwhacker.py --config /example/config.json`

**Comment by manics** [comment]
An alternative way (avoiding threads) is to have a second mode for execwhacker where it only does a batch kill of existing processes, then exits, or sleeps and re-runs after N minutes to catch any missed processes, running in a third container in the same pod.

**Review by yuvipanda — CHANGES_REQUESTED** [review]
w00t, this is awesome <3 Thanks for working on it, @manics.

Can you change this to use psutil instead? I think that'll be cleaner, and easier to reason about.

Not necessarily as part of this PR, but I think this should actually run every 10 minutes or so on the thread anyway, so we catch things that are missed by ebpf.

File: cryptnono/templates/daemonset.yaml [hunk 1]
```diff
@@ -142,6 +145,9 @@ spec:
         - hostPath:
             path: /lib/modules
           name: modules-host
+        - hostPath:
```
[the remark this answers is not in the document]
Comment (manics) [reply cryptnono/templates/daemonset.yaml]: No, I've removed it

**Comment by manics** [comment]
I thought I might as well throw in a GH workflow test
```
Run python -mpytest -v tests/
============================= test session starts ==============================
platform linux -- Python 3.11.6, pytest-7.4.3, pluggy-1.3.0 -- /opt/hostedtoolcache/Python/3.11.6/x64/bin/python
cachedir: .pytest_cache
rootdir: /home/runner/work/cryptnono/cryptnono
collecting ... collected 3 items

tests/test_cryptnono_container.py::test_allowed PASSED                   [ 33%]
tests/test_cryptnono_container.py::test_killed PASSED                    [ 66%]
tests/test_cryptnono_container.py::test_self_changing_killed PASSED      [100%]

============================== 3 passed in 13.02s ==============================

Run docker logs cryptono
2023-12-04 23:22:50,245 Found 1 substrings to check process cmdlines for
2023-12-04 23:22:50,245 Found 1 patterns to explicitly allow
2023-12-04 23:22:50,245 Compiling and loading BPF program...
2023-12-04 23:22:53,799 Took 3.55s to startup
2023-12-04 23:22:53,799 Checking existing processes
2023-12-04 23:22:53,800 Watching for processes we don't like...
2023-12-04 23:22:53,812 Killed 0 existing processes
2023-12-04 23:23:00,809 action:spared pid:5793 cmdline:/opt/hostedtoolcache/python/3.11.6/x64/bin/python -c 'print('"'"'allowed cryptnono.banned.string1'"'"')' matched:cryptnono.banned.string1 allowed-by:.*allowed\s+cryptnono\.banned\.string1 source:execwhacker.bpf
2023-12-04 23:23:00,820 action:killed pid:5794 cmdline:/opt/hostedtoolcache/python/3.11.6/x64/bin/python -c 'print('"'"'cryptnono.banned.string1'"'"')' matched:cryptnono.banned.string1 source:execwhacker.bpf
2023-12-04 23:23:03,812 Checking existing processes
2023-12-04 23:23:03,817 Killed 0 existing processes
2023-12-04 23:23:13,817 Checking existing processes
2023-12-04 23:23:13,823 action:killed pid:5795 cmdline:cryptnono.banned.string1 no/cryptnono/tests/resources/cryptnono-test-self-changing-cmdline matched:cryptnono.banned.string1 source:psutil.process_iter
2023-12-04 23:23:13,823 Killed 1 existing processes
```

File: tests/resources/cryptnono-test-self-changing-cmdline.c [hunk 2]
```diff
@@ -0,0 +1,14 @@
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+
+int main (int argc, char* argv[])
+{
+    printf("pid: %i\n", getpid());
+    sleep(5);
+    snprintf(argv[0], strlen(argv[0]), "cryptnono.banned.string1");
```
Comment (yuvipanda) [inline tests/resources/cryptnono-test-self-changing-cmdline.c:9]: Can you write a comment here on what this is doing?

File: scripts/execwhacker.py [hunk 3]
```diff
@@ -96,6 +115,24 @@ def process_event(
             logging.exception(e)
 
 
+def check_existing_processes(banned_strings_automaton, allowed_patterns, interval):
```
Comment (yuvipanda) [inline scripts/execwhacker.py:118]: Add a docstring?

File: scripts/execwhacker.py [hunk 4]
```diff
@@ -96,6 +115,24 @@ def process_event(
...
+            if proc.exe():
+                if kill_if_needed(
+                    banned_strings_automaton,
+                    allowed_patterns,
+                    join(proc.cmdline()),
+                    proc.pid,
+                    ProcessSource.SCAN,
+                ):
+                    count += 1
+        logging.info(f"Killed {count} existing processes")
```
Comment (yuvipanda) [inline scripts/execwhacker.py:132]: Let's keep it in the style of the other log messages, with a `source:`, `action:` etc to make long term grepping easier?

**Review by yuvipanda — COMMENTED** [review]
This is awesome! Left a couple of comments but happy to merge once those are fixed.

**Comment by manics** [comment]
Updated!

File: tests/resources/cryptnono-test-self-changing-cmdline.c [hunk 5]
```diff
@@ -2,13 +2,26 @@
...
+/**
+ * This provides an executable to test whether execwhacker detects processes that
+ * were missed by BPF.
+ *
+ * execwhacker should be run with arguments `--config /example/config.json --scan-existing 10`
+ *
+ * This program should be compiled to an executable named `cryptnono-test-self-changing-cmdline`
+ * which means it should be allowed by execwhacker.
+ *
+ * Afger a few seconds the executable overwrites it's argv[0] to the banned string
```
Comment (yuvipanda) [inline tests/resources/cryptnono-test-self-changing-cmdline.c:14]:
```suggestion
 * After a few seconds the executable overwrites it's argv[0] to the banned string
```

**Comment by yuvipanda** [comment]
Yay, thank you @manics. Think you can deploy this?

**Comment by manics** [comment]
https://github.com/yuvipanda/cryptnono/pull/12 (hopefully) means this will be automatically deployed by a mybinder.org-deploy workflow

**Comment by manics** [comment]
Seems to be working!

**Comment by yuvipanda** [comment]
yay, THANK YOU SO MUCH @manics

**Pull request merged**